vignettes/wrapping-apis.Rmd

```diff
@@ -213,92 +224,130 @@ faker_person <- function(gender = NULL, birthday_start = NULL, birthday_end = NU
+    ```{r}
+    secret <- secret_encrypt("this is a secret message", key)
+    secret
+    ```
 
-Now you need to make this key available in the places were you want to unlock the secrets.
-httr2 provides tools to retrieve keys from environment variables because this is a common way of communicating these sort of secrets.
-So start by opening your `.Renviron` (e.g. with `usethis::edit_r_environ()`) and add a line like:
+3.  When needed, you descramble the the secret using `secret_decrypt()`:
+
+    ```{r}
+    secret_decrypt(secret, key)
+    ```
 
-    YOURPACKAGE_KEY=key_that_you_generated_above
+### Package secrets
 
-Another place you'll probably want to provide the key is in your Github actions.
-This is a two step process:
+You can create an number of encryption keys, but I highly recommend that you create one key per package, which I'll call the **package** key.
+In this section, I'll show you how to store that key so that you (and your automated tests) can use it, but no one else can.
+
+httr2 is built around the notion that this key should live in an environment variable.
+So the first step is to make your package key available on your local development machine by adding a line to your your user-level `.Renviron` (which you can easily open with `usethis::edit_r_environ()`):
+
+    YOURPACKAGE_KEY=key_you_generated_with_secret_make_key
+
+```{r, include = FALSE}
+Sys.setenv(YOURPACKAGE_KEY = secret_make_key())
+```
+
+Now (after you restart R), you'll be able to take advantage of a special `secret_encrypt()` and `secret_decrypt()` feature: the encryption key can be the name of an environment variable.
+
+```{r}
+secret <- secret_encrypt("my personal secret", "YOURPACKAGE_KEY")
+secret
+secret_decrypt(secret, "YOURPACKAGE_KEY")
+```
+
+You'll also need to make the key available in your GitHub Actions so your automated tests can use it.
+This requires two steps:
 
 1.  Add the key to your [repository secrets](https://docs.github.com/en/actions/reference/encrypted-secrets).
 
 2.  Share the secret with the workflows that need it by adding a line to the appropriate workflow:
 
     ``` yaml
         env:
-          HTTR2_KEY: ${{ secrets.HTTR2_KEY }}
+          YOURPACKAGE_KEY: ${{ secrets.YOURPACKAGE_KEY }}
     ```
 
-    You can see how httr2 does it in <https://github.com/r-lib/httr2/blob/master/.github/workflows/R-CMD-check.yaml>.
+    You can see how httr2 does it in [its GitHub workflow](https://github.com/r-lib/httr2/blob/master/.github/workflows/R-CMD-check.yaml){.uri}.
 
-Now you can encrypt your API key and store it in your tests --- there's no way for anyone to decrypt it unless they have your key.
+Other continuous integration platforms will offer similar ways to make secrets available in environment variables.
 
-```{r}
-x <- secret_encrypt("my-secret", "HTTR2_KEY")
-x
-secret_decrypt(x, "HTTR2_KEY")
+### When the package key isn't available
+
+There's an important computer that won't have your package key: CRAN.
```

Comment:
Is it also worth mentioning that this applies to PRs from external forks? If you don't do this, you will never get any decent info from CI for such PRs. Whereas with this, assuming you have some tests that don't require live API access, at least those tests will provide info and the others will be skipped vs. failing.

Reply:
Words I have used for this topic in https://gargle.r-lib.org/articles/articles/managing-tokens-securely.html

> Decryption is not going to work. Either because the Suggested sodium package is not available or (much more likely) because the environment variable that represents the key is not available.
> This will be the case on CRAN, by definition, because there is no way to share an encrypted secret.
> This will be the case for external contributors, on their personal machines and when their GitHub pull requests are checked via CI services, such as GitHub Actions, Travis-CI, or AppVeyor.
> ...
> By calling the skipper at the start of tests that require auth, you arrange for your package to cope gracefully when the token cannot be decrypted, e.g., on CRAN and in pull requests.